### ISSUE TITLE
event-bus-event-publish-service slow to cache new channels / subscriptions

### ISSUE THREAD

**Issue by tminke** [kyma-project/kyma#6160]
**Description**
When creating new Kyma Subscriptions we see problems with the event-bus-event-publish-service pods being able to find and send to the actual Channel.  They seem to take a long time (many many minutes) to load the new Channel into their cache and/or to detect the Channel's READY state and subscriptions aren't detected.

This is being seen on Kyma 1.6 running in an Azure K8S cluster (v1.14) using a custom Kafka Channel implementation.

**Expected result**
Shortly (seconds) after a new Kyma Subscription results in Knative Kafka Channels being spun up Kyma should be able to forward messages/events to the Channel.

**Actual result**
It takes 10s of minutes for one or both of the event-bus-event-publish-service pods to detect and be able to send to the Channel.

**Steps to reproduce**
1. Tail the logs of the event-bus-event-publish-service pods
2. Create a new Kyma Subscription (ideally for the latest konduit kafka channel type)
3. Once the Channel pods are up, attempt to send events to Kyma (channel)
4. Notice in the event-bus-event-publish-service logs that it doesn't know about the channel or thinks the channel is not ready or there aren't any subscriptions.

Attached is a detailed capture of the logs and attempts to send messages to Kyma as well as the relevant config yaml from the cluster.
[Kyma-EventPublish-Bug.txt](https://github.com/kyma-project/kyma/files/3784140/Kyma-EventPublish-Bug.txt)

**Troubleshooting**
Sending events over a period of 45 minutes while tailing the logs to see if the event-bus-event-publish-service ever stabilized.  Also had to reset one of the event-bus-event-publish-service pods to resolve Istio config issues (503).  Also, correlated the timestamps of the various message sending attempts against the READY state of the channels to verify the event-bus-event-publish-service cache should have been loaded.

**Comment by antoineco** [comment]
It looks like the issue we had when the controller didn't have the `watch` permission (RBAC) on channels. The initial `list` would work, but because the watch never happened, the channels cache was never updated. Let me try to reproduce.

**Comment by antoineco** [comment]
Confirmed, we're still having a permission issue:
```
E1105 12:43:06.293366       1 reflector.go:125] github.com/knative/eventing/pkg/client/informers/externalversions/factory.go:119: Failed to list *v1alpha1.Chann
el: Get https://10.0.32.1:443/apis/messaging.knative.dev/v1alpha1/channels?limit=500&resourceVersion=0: dial tcp 10.0.32.1:443: connect: connection refused
```

**Comment by antoineco** [comment]
There is actually no permission issue, all `list`, `watch` and `get` operations succeed. I'll have to dig deeper.

**Comment by antoineco** [comment]
I tried to reproduce following the example at https://github.com/kyma-project/examples/blob/master/event-subscription/service/README.md but can't reproduce. @tminke how long does it take on average for a KafkaChannel to become ready? The default timeout is 10 sec, and it can be overridden using the `channelTimeout` cmd flag.

**Comment by tminke** [comment]
So I spent some time digging into the event-bus source and correlating with the logs (below) from the attached file and have narrowed things down a little bit.  First, of interest is the event-bus-event-publish-service log which shows the long delay from the first failed attempt ot send a message at 17:05:59 to the first successful attempt at 17:31:56 (some 26 minutes later).  In the failed attempts it was unable to find the channel based on the labels.  If you look at the knative channels.messaging.... in the attached file you can see that the labels are there (though this capture was done after things started working).  This might point to problems with the knative/go-client caching of the channels?

```
2019/10/28 17:05:59 knative channels fetched []
2019/10/28 17:05:59 no channels with the map[kyma-source-id:commerce kyma-event-type:product.created kyma-event-type-version:v1] labels were found in kyma-system namespace
2019/10/28 17:05:59 failed to get the knative channel for source: 'commerce', event-type: 'product.created', event-type-version: 'v1' in namespace 'kyma-system'
error: 'channels.messaging.knative.dev "" not found':
2019/10/28 17:05:59 incrementing ignored messages counter
2019/10/28 17:05:59 publish to the knative channel: '' namespace: 'kyma-system' status: 'ignored' reason: 'Event was ignored as there are no subscriptions or consumers configured for this event'

...

2019/10/28 17:31:56 knative channels fetched [0xc4272fe9a0]
2019/10/28 17:32:57 publish to the knative channel: 'productcr-mwz4d' namespace: 'kyma-system' status: 'published' reason: 'Message successfully published to the channel'
```

...also - the Channel shows the READY transition at 16:53:57 which is 10 minutes before the first attempt to send the message, and is only 2 seconds after the Channel creationTimestamp of 16:53:55.

 I will continue to look at this but thought this further highlighting of the data in the attached txt file might help - thanks again!

**Comment by tminke** [comment]
I captured the issue again and the information is essentially the same but I've distilled the timing out for easy understanding...

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Timeline of events
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

1. Creation Times
     Kyma Subscription    creationTimestamp: "2019-11-06T16:09:05Z"
     Knative Channel      creationTimestamp: "2019-11-06T16:09:07Z"
     Kafka Channel        creationTimestamp: "2019-11-06T16:09:07Z"
     Knative Subscription creationTimestamp: "2019-11-06T16:09:16Z"

2. Ready Status Time
     Knative Channel      lastTransitionTime: "2019-11-06T16:09:15Z"
     Kafka Channel        lastTransitionTime: "2019-11-06T16:09:15Z"
     Knative Subscription lastTransitionTime: "2019-11-06T16:09:16Z"
     Kyma Subscription    last_transition_time: "2019-11-06T16:09:17Z"

3. Sending Events
     First Event (success)  "2019-11-06T16:15:08+00:00"
     Second Event (failed)  "2019-11-06T16:18:29+00:00"
     Final Event (success)  "2019-11-06T16:25:29+00:00"



~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Takeaways ...
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

1. The Channels were  READY 10 seconds after the Kyma Subscription was created
2. The Kyma Subscription was READY 12 seconds after the Kyma Subscription was created
3. Waited over 5 minutes before sending the first event
4. First event to one of the event-bus-event-publish-services was ok
5. Second event to the other event-bus-event-publish service was unable to find channel
   even though it clearly existed - likely not in the cache?
6. 7 minutes later sending event to the problematic event-bus-event-publish-service was
   successful - found the channel (cache updated?)

... see the attached file for the full logs and configuration...
[Kyma-SubscriptionController-Bug-2.txt](https://github.com/kyma-project/kyma/files/3815978/Kyma-SubscriptionController-Bug-2.txt)

**Comment by tminke** [comment]
fwiw - some semi-related information...

Deleted the Channel to try and reproduce the issue and it resulted in another / different problem.  After everything had come back up one of the publisher services was working fine, but the other one still had the old / stale channel name in it's error message...
```
2019/11/06 18:54:43 knative channels fetched [0xc420e3c000]
2019/11/06 18:54:43 ERROR: SendMessage(): could not send HTTP request: Post http://productcr-jwj4j-channel.kyma-system.svc.cluster.local: dial tcp: lookup productcr-jwj4j-channel.kyma-system.svc.cluster.local on 10.0.0.10:53: no such host
2019/11/06 18:54:43 failed to send message to the knative channel 'productcr-jwj4j' in namespace 'kyma-system'
...while the new channel name was `productcr-tscbz-channel`
```

**Comment by tminke** [comment]
Just FYI - I reproduced the issue for Nils Schmidt in our Kyma 1.6 cluster.  We then replaced the `event-bus-event-publish-service` and `event-bus-subscription-controller` Deployments in the with the 1.8 images and haven't been able to reproduce.  The plan is to upgrade to 1.8 and confirm the problem has been resolved and then we can close this issue.

```
# 1.8 image tags
image: eu.gcr.io/kyma-project/develop/subscription-controller:af0a546c
image: eu.gcr.io/kyma-project/develop/event-publish-service:af0a546c
```

**Comment by tminke** [comment]
It appears that this problem has been resolved in Kyma 1.8.  It's hard to PROVE it won't happen but for now we can't reproduce it and will close the issue.
